### ISSUE TITLE
how to save weights

### ISSUE THREAD

**Issue by Podidiving** [qubvel/segmentation_models#161]
I am using PSPNet for multiclass segmentation. Training is fine and metrics are good. In the end I am saving model using `model.save_weights(path)`
But after loading the model some weights are `nan` (to be more precise, I have `nan`s in `level1_bn` layer) Can anybody suggest solution for this problem?

**Comment by qubvel** [comment]
Is predictions differs for actual model and loaded?

**Comment by ali-masoudi** [comment]
> I am using PSPNet for multiclass segmentation. Training is fine and metrics are good. In the end I am saving model using `model.save_weights(path)`
> But after loading the model some weights are `nan` (to be more precise, I have `nan`s in `level1_bn` layer) Can anybody suggest solution for this problem?

Can you post your result, cause I'm not getting a good result on cityscapes?
I train the network for 30 epochs with frozen encoder, then unfreeze the encoder and resume training, but no problem in saving weights.
**But the saved model has no optimizer and loss**
Segmentation_models,Classification_modesl,Efficient_Net : all the latest version
**my training code**

``` python
from segmentation_models import PSPNet
from segmentation_models.metrics import IOUScore
from segmentation_models.losses import JaccardLoss
from keras.callbacks import ModelCheckpoint
from keras.callbacks import TensorBoard
from keras.callbacks import LearningRateScheduler
from keras.utils import multi_gpu_model
import os
import argparse
from CustomGenerator_il import CustomGenerator
import tensorflow as tf
from utils import PolyDecay
from keras import backend as K
from keras.optimizers import SGD
from keras.callbacks import CSVLogger
from segmentation_models.utils import set_trainable


parser = argparse.ArgumentParser()
parser.add_argument('--batch_size', type=int, default=16, help='input batch size')
parser.add_argument('--image_width', type=int, default=384, help='the input image width')
parser.add_argument('--image_height', type=int, default=384, help='the input image height')
parser.add_argument('--lr', type=float, default=0.01, help='learning rate of the optimizer')
parser.add_argument('--decay', type=float, default=0.9, help='learning rate decay (per epoch)')
parser.add_argument('--n_epochs', type=int, default=100, help='number of epochs to train for')
parser.add_argument('--n_gpu', type=int, default=1, help='number of GPUs to use')
parser.add_argument('--n_cpu', type=int, default=12, help='number of CPU threads to use during data generation')
parser.add_argument('--checkpoint', type=str, default=None, help='path to model checkpoint to resume training')
parser.add_argument('--epoch', type=int, default=0, help='epoch to start training')
opt = parser.parse_args()

print(opt)
os.environ["CUDA_VISIBLE_DEVICES"] = "0"
config = tf.ConfigProto()
config.gpu_options.allow_growth = True
sess = tf.Session(config=config)
K.set_session(sess)


checkpoint = ModelCheckpoint(filepath='output_il/epoch_{epoch:03d}--val_iou_score_{val_iou_score:.3f}--val_loss_{val_loss:.3f}.h5',
                             monitor='val_loss',
                             mode='max')
tensorboard = TensorBoard(batch_size=opt.batch_size)
lr_decay = LearningRateScheduler(PolyDecay(opt.lr, opt.decay, opt.n_epochs).scheduler)

train_generator = CustomGenerator(mode='train',
                                  batch_size=opt.batch_size,
                                  crop_shape=(opt.image_height, opt.image_width),
                                  resize_shape=None)
val_generator = CustomGenerator(mode='val',
                                batch_size=opt.batch_size,
                                crop_shape=None,
                                resize_shape=(opt.image_height,opt.image_width))

optim = SGD(lr=opt.lr, momentum=0.9)
loss=JaccardLoss(class_weights=None,class_indexes=None,per_image=False)
iou=IOUScore(class_weights=None,class_indexes=None,per_image=False)
csv_logger = CSVLogger("model_history_log.csv", append=True)

# Model
if not opt.checkpoint:  #From Start:
    net = PSPNet(backbone_name='resnet50',
                       input_shape=(opt.image_height, opt.image_width, 3),
                       classes=20,
                       activation='softmax',
                       weights=None,
                       encoder_weights='imagenet',
                       encoder_freeze=True,
                       downsample_factor=4,
                       psp_conv_filters=512,
                       psp_pooling_type='avg',
                       psp_use_batchnorm=True,
                       psp_dropout=None)



elif opt.checkpoint:    #Resume From Checkpoint
    net = PSPNet(backbone_name='resnet50',
                 input_shape=(opt.image_height, opt.image_width, 3),
                 classes=20,
                 activation='softmax',
                 weights=opt.checkpoint,
                 encoder_weights='imagenet',
                 encoder_freeze=True,
                 downsample_factor=4,
                 psp_conv_filters=512,
                 psp_pooling_type='avg',
                 psp_use_batchnorm=True,
                 psp_dropout=None)
    # OR
    # net = load_model(filepath=opt.checkpoint,
    #                  compile=False,
    #                  custom_objects={'jaccard_loss': loss})
if opt.n_gpu > 1:
    net = multi_gpu_model(net, opt.n_gpu)

# Training

if not opt.checkpoint:  #From Start
    net.compile(optim, loss, metrics=[iou])

elif opt.checkpoint:    #Resume From Checkpoint
    set_trainable(net, False)
    net.compile(optim, loss, metrics=[iou])

net.fit_generator(train_generator, steps_per_epoch=None, epochs=opt.n_epochs, callbacks=[checkpoint, tensorboard, lr_decay,csv_logger],
                    validation_data=val_generator, validation_steps=None, workers=10,
                    use_multiprocessing=True, shuffle=True, max_queue_size=10, initial_epoch=opt.epoch)

```

**Comment by qubvel** [comment]
@ali-masoudi Training PSPNet also required an auxiliary output, please refer to original paper.

**Comment by hasan-nn** [comment]
hello, I may have a relevant question,
does loading weights by setting weights=(checkpoint model) in sm.Unet(......,weights=ckpt,encoder_weights='image_net', .. ) overwrite the image net encoder weights?
